polars/polars-lazy/polars-plan/src/dsl/function_expr/schema.rs

```diff
@@ -177,9 +177,14 @@ impl FunctionExpr {
                 match s {
                     FieldByIndex(index) => {
                         let (index, _) = slice_offsets(*index, 0, fields.len());
-                        fields.get(index).cloned().ok_or_else(
-                            || polars_err!(ComputeError: "index out of bounds in `struct.field`"),
-                        )
+                        if let DataType::Struct(flds) = &fields[0].dtype {
+                            let fld = flds.get(index).cloned().ok_or_else(
+                            || polars_err!(ComputeError: "index out of bounds in `struct.field`")
+                            )?;
+                            Ok(fld)
+                        } else {
+                            polars_bail!(ComputeError: "index out of bounds in `struct.field`");
```

Comment:
I believe this should be `unreachable!`. We only get here via the `struct` namespace. 

If not, we must return an error indicating that we expected a struct.